service/history/queues/actions.go

```diff
@@ -0,0 +1,118 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type (
+	// Action is operations that can be run on a ReaderGroup.
+	// It is created by Mitigator upon receiving an Alert and
+	// run by a Queue to resolve the alert.
+	Action interface {
+		Run(*ReaderGroup)
+	}
+
+	actionReaderStuck struct {
+		mitigator *mitigatorImpl
+
+		attributes *AlertAttributesReaderStuck
+		logger     log.Logger
+	}
+)
+
+func newReaderStuckAction(
+	mitigator *mitigatorImpl,
+	attributes *AlertAttributesReaderStuck,
+	logger log.Logger,
+) Action {
+	return &actionReaderStuck{
+		mitigator:  mitigator,
+		attributes: attributes,
+		logger:     logger,
+	}
+}
+
+func (a *actionReaderStuck) Run(readerGroup *ReaderGroup) {
+	defer a.mitigator.resolve(AlertTypeReaderStuck)
+
+	if a.attributes.ReaderID == int32(a.mitigator.maxReaderCount()-1) {
+		return
+	}
```

Comment:
Monitor already has logic to skip alert if readerID != defaultReaderId.

Reply:
Make sense. Decoupled action from mitigator as you suggested.

Also removed max reader count check from this action since monitor can ensure not alert is sent. 
For other other future action, max reader count may still be required (e.g. for pending task action which is not tied to a particular reader), and it can be passed in when creating the action through newXXXAction().